content_script.js

```diff
@@ -137,38 +140,104 @@ class HTMLSerializer {
    *
    * @param {Element} element The element being processed, which has the src
    *     attribute.
+   * @param {Set<string>} processedAttributes The Set containing all attributes
+   *     already added to |this.html| for the Element.
    * @private
    */
-  processSrcAttribute(element) {
-    if (element.tagName.toLowerCase() != 'iframe') {
-      this.processSrcHole(element);
-    } 
+  processSrcAttribute(element, processedAttributes) {
+    var tag = element.tagName.toLowerCase();
+    switch(tag) {
+      case 'iframe':
+        break; // Do nothing.
+      case 'source':
+        if (!element.parent || element.parent.tagName.toLowerCase() != 'img') {
+          this.processSimpleSrc(element, processedAttributes);
+          break;
+        } // else process as img.
+      case 'input':
+        var type = element.attributes.type;
+        if (tag == 'input' && (!type || type.value.toLowerCase() != 'image')) {
+          break;
+        } // else process as img.
+      case 'img':
+        if (window.location.host == this.fullyQualifiedURL(element).host) {
+          this.processSrcHole(element, processedAttributes);
+          break;
+        }
+      default:
+        this.processSimpleSrc(element, processedAttributes);
+    }
+  }
+
+  /**
+   * Get a URL object for the value of the elements src attribute.
+   *
+   * @param {Element} element The element for which to retrieve the URL.
+   * @return {URL} The URL object.
+   */
+  fullyQualifiedURL(element) {
+    var url = element.attributes.src.value;
+    var a = document.createElement('a');
+    a.href = url;
+    url = a.href; // Retrieve fully qualified URL.
+    return new URL(url);
```

Comment:
Just because the spec says that doesn't mean that content will obey the spec. :(

How about you process the attributes in reverse order for this reason?